envision/web/src/components/simulation.js

```diff
@@ -149,18 +149,26 @@ export default function Simulation({
     setScene(scene_);
   };
 
+  const sleep = (milliseconds) => {
+    return new Promise((resolve) => setTimeout(resolve, milliseconds));
+  };
+
   // State subscription
   useEffect(() => {
     let stopPolling = false;
     (async () => {
       const it = client.worldstate(simulationId);
-      let wstate_and_time = await it.next();
-      while (!wstate_and_time.done && playing) {
+      let wstate_and_time;
+      if (playing) wstate_and_time = await it.next();
+      while (!stopPolling && playing && !wstate_and_time.done) {
         let wstate, elapsed_times;
         [wstate, elapsed_times] = wstate_and_time.value;
         if (!stopPolling) {
           setWorldState(wstate);
           onElapsedTimesChanged(...elapsed_times);
+
+          // play the frames at 100/40 = 20fps
+          await sleep(50);
```

Comment:
This seems suspicious to me, this calculation wants `100/50` but since this waits in real-time what stops this from playing at ~200fps?

Reply:
this line causes the code execution to sleep for 50ms before fetching the next `wstate_and_time`.

My comment was a bit incorrect, it should be 1s = 1000ms/50ms = 20 frames per second. This ensures the simulation doesn't run faster than 20 frames per second, 20 fps also seems to be the normal rate for smarts to send out the frames.

I updated the comment and this 50ms here is also open to changes.